File: src/main/java/no/nav/aareg/teknisk/historikk/KonsumentLogging.java
package no.nav.aareg.teknisk.historikk;

import io.micrometer.core.instrument.MeterRegistry;
import jakarta.servlet.http.HttpServletRequest;

public final class KonsumentLogging {

    public static final String AAREG_ARBEIDSFORHOLD_OPPSLAG_NAVN = "aareg_arbeidsforhold_oppslag";
    public static final String MDC_CONSUMER_ID = "consumerId";
    public static final String MDC_SUPPLIER_ID = "supplierId";
    public static final String OPPSLAGSTYPE_METRIC_TAG_NAVN = "oppslagstype";
    public static final String ORG_NUMMER_METRIC_TAG_NAVN = "organisasjon";
    public static final String DATABEHANDLER_METRIC_TAG_NAVN = "databehandler";

    private KonsumentLogging() {
    }

    public static void loggOppslag(MeterRegistry meterRegistry, HttpServletRequest request, String oppslagsnavn, int size) {
        meterRegistry.counter(
                AAREG_ARBEIDSFORHOLD_OPPSLAG_NAVN,
                OPPSLAGSTYPE_METRIC_TAG_NAVN, oppslagsnavn,
                ORG_NUMMER_METRIC_TAG_NAVN, (String) request.getAttribute(MDC_CONSUMER_ID),
                DATABEHANDLER_METRIC_TAG_NAVN, (String) java.util.Objects.requireNonNullElse(request.getAttribute(MDC_SUPPLIER_ID), "")
        ).increment(size);
    }
}

File: src/main/kotlin/no/nav/aareg/teknisk_historikk/KonsumentLogging.kt
package no.nav.aareg.teknisk_historikk

import io.micrometer.core.instrument.MeterRegistry
import org.slf4j.MDC
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

const val AAREG_ARBEIDSFORHOLD_OPPSLAG_NAVN = "aareg_arbeidsforhold_oppslag"
const val MDC_CONSUMER_ID = "consumerId"
const val MDC_SUPPLIER_ID = "supplierId"

const val OPPSLAGSTYPE_METRIC_TAG_NAVN = "oppslagstype"
const val ORG_NUMMER_METRIC_TAG_NAVN = "organisasjon"
const val DATABEHANDLER_METRIC_TAG_NAVN = "databehandler"

@Component
class KonsumentFilter : OncePerRequestFilter() {
    override fun doFilterInternal(req: HttpServletRequest, response: HttpServletResponse, filterChain: FilterChain) {
        try {
            val orgNummer = hentOrgnrFraToken()

            MDC.put(MDC_CONSUMER_ID, orgNummer.konsument)
            if (orgNummer.databehandler != null) MDC.put(MDC_SUPPLIER_ID, orgNummer.databehandler)

            req.setAttribute(MDC_CONSUMER_ID, orgNummer.konsument)
            req.setAttribute(MDC_SUPPLIER_ID, orgNummer.databehandler)

            filterChain.doFilter(req, response)
        } finally {
            MDC.remove(KORRELASJONSID_HEADER)
        }
    }
}

fun MeterRegistry.loggOppslag(request: HttpServletRequest, oppslagsnavn: String, size: Int) {
    this.counter(
        AAREG_ARBEIDSFORHOLD_OPPSLAG_NAVN,
        OPPSLAGSTYPE_METRIC_TAG_NAVN, oppslagsnavn,
        ORG_NUMMER_METRIC_TAG_NAVN, request.getAttribute(MDC_CONSUMER_ID) as String,
        DATABEHANDLER_METRIC_TAG_NAVN, (request.getAttribute(MDC_SUPPLIER_ID) as String?) ?: ""
    )
        .increment(size.toDouble())
}